runtime/lua/vim/diagnostic.lua

```diff
@@ -2176,44 +2299,81 @@ function M.open_float(opts, ...)
     opts = get_resolved_options({ float = float_opts }, nil, bufnr).float
   end
 
+  local buf_lines = api.nvim_buf_get_lines(bufnr, 0, -1, true)
   local scope = ({ l = 'line', c = 'cursor', b = 'buffer' })[opts.scope] or opts.scope or 'line'
-  local lnum, col --- @type integer, integer
-  local opts_pos = opts.pos
+  local extmarks_start = { 0, 0 }
+  local extmarks_end = { -1, -1 }
   if scope == 'line' or scope == 'cursor' then
+    local opts_pos = opts.pos
+
     if not opts_pos then
       local pos = api.nvim_win_get_cursor(0)
-      lnum = pos[1] - 1
-      col = pos[2]
+      extmarks_start[1] = pos[1] - 1
+
+      if scope == 'cursor' then
+        extmarks_start[2] = pos[2]
+      end
     elseif type(opts_pos) == 'number' then
-      lnum = opts_pos
+      extmarks_start[1] = opts_pos
     elseif type(opts_pos) == 'table' then
-      lnum, col = opts_pos[1], opts_pos[2]
+      extmarks_start[1] = opts_pos[1]
+
+      if scope == 'cursor' then
+        extmarks_start[2] = opts_pos[2]
+      end
     else
       error("Invalid value for option 'pos'")
     end
+
+    extmarks_end[1] = extmarks_start[1]
+
+    if scope == 'cursor' then
+      extmarks_start[2] = math.min(extmarks_start[2], #buf_lines[extmarks_start[1] + 1] - 1)
+      extmarks_end[2] = extmarks_start[2]
+    end
   elseif scope ~= 'buffer' then
     error("Invalid value for option 'scope'")
   end
 
-  local diagnostics = get_diagnostics(bufnr, opts --[[@as vim.diagnostic.GetOpts]], true)
+  local namespaces ---@type integer[]
+  local opts_namespace = opts.namespace
+  if type(opts_namespace) == 'number' then
+    namespaces = { opts_namespace }
+  elseif type(opts_namespace) == 'table' then
+    namespaces = opts_namespace
+  else
+    namespaces = vim.tbl_keys(diagnostic_cache[bufnr])
+  end
 
-  if scope == 'line' then
-    --- @param d vim.Diagnostic
-    diagnostics = vim.tbl_filter(function(d)
-      return lnum >= d.lnum
-        and lnum <= d.end_lnum
-        and (d.lnum == d.end_lnum or lnum ~= d.end_lnum or d.end_col ~= 0)
-    end, diagnostics)
-  elseif scope == 'cursor' then
-    -- If `col` is past the end of the line, show if the cursor is on the last char in the line
-    local line_length = #api.nvim_buf_get_lines(bufnr, lnum, lnum + 1, true)[1]
-    --- @param d vim.Diagnostic
-    diagnostics = vim.tbl_filter(function(d)
-      return lnum >= d.lnum
-        and lnum <= d.end_lnum
-        and (lnum ~= d.lnum or col >= math.min(d.col, line_length - 1))
-        and ((d.lnum == d.end_lnum and d.col == d.end_col) or lnum ~= d.end_lnum or col < d.end_col)
-    end, diagnostics)
+  ---@type vim.Diagnostic[]
+  local diagnostics = {}
+  for _, namespace in ipairs(namespaces) do
```

Comment:
same here: 1-line comment that gives an intuition of what this nested for-loop is doing.

Reply:
added a comment before the loop and inside